standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java

```diff
@@ -2174,86 +2270,132 @@ private List<ColumnStatisticsObj> aggrStatsUseDB(String catName, String dbName,
+                } else {
+                  row[2 + colStatIndex] = extrapolateMethod.extrapolate(min, max, colStatIndex, indexMap);
                 }
-                fqr = (ForwardQueryResult<?>) qResult;
-                Object[] avg = (Object[]) (fqr.get(0));
-                // colStatIndex=12,13,14 respond to "AVG_LONG", "AVG_DOUBLE",
-                // "AVG_DECIMAL"
-                row[2 + colStatIndex] = avg[colStatIndex - 12];
-                end = doTrace ? System.nanoTime() : 0;
-                MetastoreDirectSqlUtils.timingTrace(doTrace, queryText, start, end);
+              } finally {
+                columnWisePartitionBatches.closeAllQueries();
               }
+            } else {
+              // colStatIndex=12,13,14 respond to "AVG_LONG", "AVG_DOUBLE",
+              // "AVG_DECIMAL"
+              row[2 + colStatIndex] = avg[colStatIndex - 12];
             }
           }
-          colStats.add(prepareCSObjWithAdjustedNDV(row, 0, useDensityFunctionForNDVEstimation, ndvTuner));
+          colStats.add(columnStatisticsObjWithAdjustedNDV
+                  (Collections.singletonList(row), useDensityFunctionForNDVEstimation, ndvTuner));
           Deadline.checkTimeout();
         }
       }
       return colStats;
     }
   }
 
+  private ColumnStatisticsObj columnStatisticsObjWithAdjustedNDV(
+          List<Object[]> columnBatchesOutput,
+          boolean useDensityFunctionForNDVEstimation, double ndvTuner)
+          throws MetaException {
+    if (columnBatchesOutput.isEmpty()) {
+      return null;
+    }
+    ColumnStatisticsData data = new ColumnStatisticsData();
+    Object[] row = columnBatchesOutput.getFirst();
+    String colName = (String) row[COLNAME];
+    String colType = (String) row[COLTYPE];
+    ColumnStatisticsObj cso = new ColumnStatisticsObj(colName, colType, data);
+    Object llow = row[LLOW];
+    Object lhigh = row[LHIGH];
+    Object dlow = row[DLOW];
+    Object dhigh = row[DHIGH];
+    Object declow = row[DECLOW];
+    Object dechigh = row[DECHIGH];
+    Object nulls = row[NULLS];
+    Object dist = row[DIST];
+    Object avglen = row[AVGLEN];
+    Object maxlen = row[MAXLEN];
+    Object trues = row[TRUES];
+    Object falses = row[FALSES];
+    Object avgLong = row[AVGLONG];
+    Object avgDouble = row[AVGDOUBLE];
+    Object avgDecimal = row[AVGDECIMAL];
+    Object sumDist = row[SUMDIST];
+    if (row.length == 18) {
+      StatObjectConverter.fillColumnStatisticsData(cso.getColType(), data, llow, lhigh, dlow, dhigh, declow, dechigh,
+              nulls, dist, avglen, maxlen, trues, falses, avgLong, avgDouble, avgDecimal, sumDist,
+              useDensityFunctionForNDVEstimation, ndvTuner);
+      return cso;
+    }
+    Object sumLong = row[AVGLONG];
+    Object countLong = row[AVGLONG + 1];
+    Object sumDouble = row[AVGDOUBLE + 1];
+    Object countDouble = row[AVGDOUBLE + 2];
+    Object sumDecimal = row[AVGDECIMAL + 2];
+    Object countDecimal = row[AVGDECIMAL + 3];
```

Comment:
Why are the constants used for indexing called `AVG...`? Aren't these `sums` and `counts`?